binary_search_tree/tree.py

```diff
@@ -1,48 +1,115 @@
-class TreeNode:
-    def __init__(self, key, val = None):
-        if val == None:
-            val = key
+
+class TreeNode: 
+    def __init__(self, key, value = None):
+        if value == None:
+            value = key
 
         self.key = key
-        self.value = val
+        self.value = value
         self.left = None
         self.right = None
-        
-
 
 class Tree:
     def __init__(self):
         self.root = None
+    
+    def add_helper(self, current, key, value):
+        if not current:
+            return TreeNode(key,value)
+        if key <= current.key:
+            current.left = self.add_helper(current.left,key,value)
+        else:
+            current.right = self.add_helper(current.right,key,value)
+        return current
 
-    # Time Complexity: 
-    # Space Complexity: 
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key,value)
+        else:
+            self.add_helper(self.root, key,value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+        
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find(self, key):
```

Comment:
👍 However the space complexity is O(1) because it's an iterative solution (no call stack).

The time complexity is O(log n) if the tree is balanced and O(n) worst-case.